speaker/bgp_controller.go

```diff
@@ -272,20 +447,301 @@ type session interface {
+				}
+			case pad.FromLabels.Port:
+				port, err := strconv.ParseUint(v, 10, 16)
+				if err != nil {
+					return nil, fmt.Errorf("parsing peer port: %v", err)
+				}
+				peerPort = uint16(port)
+			case pad.FromLabels.SrcAddr:
+				srcAddr = net.ParseIP(v)
+				if v != "" && srcAddr == nil {
+					return nil, fmt.Errorf("invalid source IP %q", v)
+				}
+			case pad.FromLabels.HoldTime:
+				holdTimeRaw = v
+			case pad.FromLabels.RouterID:
+				routerID = net.ParseIP(v)
+				if routerID == nil {
+					return nil, fmt.Errorf("invalid router ID %q", v)
+				}
+			}
+		}
+	}
+
+	if pad.FromAnnotations != nil {
+		for k, v := range anns {
+			switch k {
+			case pad.FromAnnotations.MyASN:
+				asn, err := strconv.ParseUint(v, 10, 32)
+				if err != nil {
+					return nil, fmt.Errorf("parsing local ASN: %v", err)
+				}
+				myASN = uint32(asn)
+			case pad.FromAnnotations.ASN:
+				asn, err := strconv.ParseUint(v, 10, 32)
+				if err != nil {
+					return nil, fmt.Errorf("parsing peer ASN: %v", err)
+				}
+				peerASN = uint32(asn)
+			case pad.FromAnnotations.Addr:
+				peerAddr = net.ParseIP(v)
+				if peerAddr == nil {
+					return nil, fmt.Errorf("invalid peer IP %q", v)
+				}
+			case pad.FromAnnotations.Port:
+				port, err := strconv.ParseUint(v, 10, 16)
+				if err != nil {
+					return nil, fmt.Errorf("parsing peer port: %v", err)
+				}
+				peerPort = uint16(port)
+			case pad.FromAnnotations.SrcAddr:
+				srcAddr = net.ParseIP(v)
+				if v != "" && srcAddr == nil {
+					return nil, fmt.Errorf("invalid source IP %q", v)
+				}
+			case pad.FromAnnotations.HoldTime:
+				holdTimeRaw = v
+			case pad.FromAnnotations.RouterID:
+				routerID = net.ParseIP(v)
+				if routerID == nil {
+					return nil, fmt.Errorf("invalid router ID %q", v)
+				}
+			}
+		}
+	}
+
+	if myASN == 0 {
+		return nil, errors.New("missing or invalid local ASN")
+	}
+	if peerASN == 0 {
+		return nil, errors.New("missing or invalid peer ASN")
+	}
+	if peerAddr == nil {
+		return nil, errors.New("missing or invalid peer address")
+	}
+
+	if holdTimeRaw != "" {
+		ht, err := config.ParseHoldTime(holdTimeRaw)
+		if err != nil {
+			return nil, fmt.Errorf("parsing hold time: %v", err)
+		}
+		holdTime = ht
+	}
+
+	// The peer is configured on a specific node object, so we want to create a
+	// BGP session only on that node.
+	h := ls[v1.LabelHostname]
+	if h == "" {
+		return nil, fmt.Errorf("label %s not found on node", v1.LabelHostname)
+	}
+	ns, err := labels.Parse(fmt.Sprintf("%s=%s", v1.LabelHostname, h))
+	if err != nil {
+		return nil, fmt.Errorf("parsing node selector: %v", err)
+	}
+
+	p := &peer{
+		Cfg: &config.Peer{
+			MyASN:         myASN,
+			ASN:           peerASN,
+			Addr:          peerAddr,
+			Port:          peerPort,
+			SrcAddr:       srcAddr,
+			HoldTime:      holdTime,
+			RouterID:      routerID,
+			NodeSelectors: []labels.Selector{ns},
+			// BGP passwords aren't supported for node peers.
```

Comment:
Maybe file a follow-up bug to track this?

Reply:
See https://github.com/metallb/metallb/issues/529#issue-566518744 and https://github.com/metallb/metallb/issues/408. Feel free to open another issue if you think it's useful.